# Comparing `tmp/pciSeq-0.0.50-py3-none-any.whl.zip` & `tmp/pciSeq-0.0.51-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,49 +1,49 @@
-Zip file size: 353530 bytes, number of entries: 67
--rw-rw-rw-  2.0 fat     2086 b- defN 23-Apr-26 22:13 pciSeq/__init__.py
--rw-rw-rw-  2.0 fat     9326 b- defN 23-May-27 20:49 pciSeq/app.py
--rw-rw-rw-  2.0 fat     3333 b- defN 23-May-27 19:24 pciSeq/config.py
+Zip file size: 354695 bytes, number of entries: 67
+-rw-rw-rw-  2.0 fat     2086 b- defN 23-May-28 20:39 pciSeq/__init__.py
+-rw-rw-rw-  2.0 fat     9252 b- defN 23-May-29 22:10 pciSeq/app.py
+-rw-rw-rw-  2.0 fat     3333 b- defN 23-May-28 20:42 pciSeq/config.py
 -rw-rw-rw-  2.0 fat     3678 b- defN 22-Apr-07 23:15 pciSeq/make_plot.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/__init__.py
--rw-rw-rw-  2.0 fat       22 b- defN 23-May-27 20:49 pciSeq/src/_version.py
+-rw-rw-rw-  2.0 fat       22 b- defN 23-May-29 22:10 pciSeq/src/_version.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/cell_call/__init__.py
--rw-rw-rw-  2.0 fat    20581 b- defN 23-May-27 20:49 pciSeq/src/cell_call/datatypes.py
+-rw-rw-rw-  2.0 fat    20857 b- defN 23-May-29 22:10 pciSeq/src/cell_call/datatypes.py
 -rw-rw-rw-  2.0 fat     2995 b- defN 23-Apr-26 22:13 pciSeq/src/cell_call/log_config.py
--rw-rw-rw-  2.0 fat    13927 b- defN 23-May-27 20:49 pciSeq/src/cell_call/main.py
--rw-rw-rw-  2.0 fat     3104 b- defN 23-May-27 20:49 pciSeq/src/cell_call/summary.py
--rw-rw-rw-  2.0 fat     8937 b- defN 23-May-27 20:49 pciSeq/src/cell_call/utils.py
+-rw-rw-rw-  2.0 fat    16544 b- defN 23-May-29 22:10 pciSeq/src/cell_call/main.py
+-rw-rw-rw-  2.0 fat     3104 b- defN 23-May-28 20:42 pciSeq/src/cell_call/summary.py
+-rw-rw-rw-  2.0 fat     8937 b- defN 23-May-28 20:42 pciSeq/src/cell_call/utils.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/preprocess/__init__.py
--rw-rw-rw-  2.0 fat    11253 b- defN 23-May-27 20:49 pciSeq/src/preprocess/cell_borders.py
+-rw-rw-rw-  2.0 fat    11253 b- defN 23-May-28 20:42 pciSeq/src/preprocess/cell_borders.py
 -rw-rw-rw-  2.0 fat     6973 b- defN 23-Apr-26 22:13 pciSeq/src/preprocess/segmentation.py
--rw-rw-rw-  2.0 fat     4090 b- defN 23-May-27 20:49 pciSeq/src/preprocess/spot_labels.py
--rw-rw-rw-  2.0 fat       88 b- defN 23-May-27 09:58 pciSeq/src/preprocess/utils.py
+-rw-rw-rw-  2.0 fat     4585 b- defN 23-May-29 22:10 pciSeq/src/preprocess/spot_labels.py
+-rw-rw-rw-  2.0 fat       88 b- defN 23-May-28 20:42 pciSeq/src/preprocess/utils.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Feb-20 14:12 pciSeq/src/viewer/__init__.py
--rw-rw-rw-  2.0 fat     2256 b- defN 23-May-27 09:58 pciSeq/src/viewer/run_flask.py
+-rw-rw-rw-  2.0 fat     2256 b- defN 23-May-28 20:42 pciSeq/src/viewer/run_flask.py
 -rw-rw-rw-  2.0 fat     4511 b- defN 23-Apr-26 22:13 pciSeq/src/viewer/stage_image.py
--rw-rw-rw-  2.0 fat     7638 b- defN 23-May-27 20:49 pciSeq/src/viewer/utils.py
--rw-rw-rw-  2.0 fat    10895 b- defN 23-May-27 09:58 pciSeq/static/2D/index.html
+-rw-rw-rw-  2.0 fat     7638 b- defN 23-May-28 20:42 pciSeq/src/viewer/utils.py
+-rw-rw-rw-  2.0 fat    10895 b- defN 23-May-28 20:42 pciSeq/static/2D/index.html
 -rw-rw-rw-  2.0 fat     6506 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/genes_datatable.html
 -rw-rw-rw-  2.0 fat     9622 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/css/index.css
--rw-rw-rw-  2.0 fat     3135 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/css/progress.css
--rw-rw-rw-  2.0 fat     7435 b- defN 23-May-27 20:48 pciSeq/static/2D/viewer/js/classConfig.js
--rw-rw-rw-  2.0 fat     1211 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/config.js
+-rw-rw-rw-  2.0 fat     3135 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/css/progress.css
+-rw-rw-rw-  2.0 fat     7435 b- defN 23-May-29 22:10 pciSeq/static/2D/viewer/js/classConfig.js
+-rw-rw-rw-  2.0 fat     1211 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/config.js
 -rw-rw-rw-  2.0 fat     5589 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/customControl.js
--rw-rw-rw-  2.0 fat    22840 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/dapi.js
+-rw-rw-rw-  2.0 fat    22840 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/dapi.js
 -rw-rw-rw-  2.0 fat     5984 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/dataLoader.js
 -rw-rw-rw-  2.0 fat    17232 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/donut.js
 -rw-rw-rw-  2.0 fat     4106 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/dt.js
--rw-rw-rw-  2.0 fat     9835 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/glyphConfig.js
+-rw-rw-rw-  2.0 fat     9835 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/glyphConfig.js
 -rw-rw-rw-  2.0 fat     4389 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/glyphPaths.js
 -rw-rw-rw-  2.0 fat     7224 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/glyphs.js
--rw-rw-rw-  2.0 fat    12281 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/index.js
+-rw-rw-rw-  2.0 fat    12281 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/index.js
 -rw-rw-rw-  2.0 fat     1073 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/progress.js
 -rw-rw-rw-  2.0 fat     6367 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/stage_cells.js
--rw-rw-rw-  2.0 fat    12064 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/stage_glyphs.js
+-rw-rw-rw-  2.0 fat    12064 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/stage_glyphs.js
 -rw-rw-rw-  2.0 fat     9540 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/stage_markers.js
--rw-rw-rw-  2.0 fat     7150 b- defN 23-May-27 09:58 pciSeq/static/2D/viewer/js/stage_markers_patched.js
+-rw-rw-rw-  2.0 fat     7150 b- defN 23-May-28 20:42 pciSeq/static/2D/viewer/js/stage_markers_patched.js
 -rw-rw-rw-  2.0 fat    15523 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/stage_polygons.js
 -rw-rw-rw-  2.0 fat     4471 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/streaming-tsv-parser.js
 -rw-rw-rw-  2.0 fat     4692 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/viewerUtils.js
 -rw-rw-rw-  2.0 fat      995 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/css/L.Control.Layers.Tree.css
 -rw-rw-rw-  2.0 fat      421 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/css/Leaflet.Coordinates-0.1.3.css
 -rw-rw-rw-  2.0 fat   122544 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/css/bootstrap.min.css
 -rw-rw-rw-  2.0 fat     8669 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/css/index.css
@@ -56,14 +56,14 @@
 -rw-rw-rw-  2.0 fat    15374 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/leaflet-pip.js
 -rw-rw-rw-  2.0 fat     5927 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/leaflet.textpath.js
 -rw-rw-rw-  2.0 fat      265 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/preloader.js
 -rw-rw-rw-  2.0 fat     9097 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/MarkerContainer.js
 -rw-rw-rw-  2.0 fat     4434 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/bezier-easing.js
 -rw-rw-rw-  2.0 fat   613808 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js
 -rw-rw-rw-  2.0 fat    91874 b- defN 23-Apr-26 22:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js
--rw-rw-rw-  2.0 fat     1091 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/LICENCE
--rw-rw-rw-  2.0 fat     3697 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/METADATA
--rw-rw-rw-  2.0 fat       92 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/WHEEL
--rw-rw-rw-  2.0 fat       48 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/entry_points.txt
--rw-rw-rw-  2.0 fat        7 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/top_level.txt
-?rw-rw-r--  2.0 fat     6383 b- defN 23-May-27 20:50 pciSeq-0.0.50.dist-info/RECORD
-67 files, 1295157 bytes uncompressed, 343164 bytes compressed:  73.5%
+-rw-rw-rw-  2.0 fat     1091 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/LICENCE
+-rw-rw-rw-  2.0 fat     4090 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/METADATA
+-rw-rw-rw-  2.0 fat       92 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/WHEEL
+-rw-rw-rw-  2.0 fat       48 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/entry_points.txt
+-rw-rw-rw-  2.0 fat        7 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/top_level.txt
+?rw-rw-r--  2.0 fat     6383 b- defN 23-May-29 22:13 pciSeq-0.0.51.dist-info/RECORD
+67 files, 1298864 bytes uncompressed, 344329 bytes compressed:  73.5%
```

## zipnote {}

```diff
@@ -177,26 +177,26 @@
 
 Filename: pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js
 Comment: 
 
 Filename: pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/LICENCE
+Filename: pciSeq-0.0.51.dist-info/LICENCE
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/METADATA
+Filename: pciSeq-0.0.51.dist-info/METADATA
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/WHEEL
+Filename: pciSeq-0.0.51.dist-info/WHEEL
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/entry_points.txt
+Filename: pciSeq-0.0.51.dist-info/entry_points.txt
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/top_level.txt
+Filename: pciSeq-0.0.51.dist-info/top_level.txt
 Comment: 
 
-Filename: pciSeq-0.0.50.dist-info/RECORD
+Filename: pciSeq-0.0.51.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## pciSeq/app.py

```diff
@@ -226,10 +226,9 @@
     _scRNAseq = pd.read_csv(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'scRNA', 'scRNAseq.csv.gz'),
                             header=None, index_col=0, compression='gzip', dtype=object)
     _scRNAseq = _scRNAseq.rename(columns=_scRNAseq.iloc[0], copy=False).iloc[1:]
     _scRNAseq = _scRNAseq.astype(float).astype(np.uint32)
 
     # main task
     # _opts = {'max_iter': 10}
-    fit(spots=_iss_spots, coo=_coo, scRNAseq=None, opts={'save_data': True,
-                                                         'launch_viewer': True,})
+    fit(spots=_iss_spots, coo=_coo, opts={'save_data': True,'launch_viewer': True,})
```

## pciSeq/src/_version.py

```diff
@@ -1 +1 @@
-__version__ = '0.0.50'
+__version__ = '0.0.51'
```

## pciSeq/src/cell_call/datatypes.py

```diff
@@ -1,12 +1,13 @@
 import sys
 import numpy as np
 import pandas as pd
 import scipy
 import numexpr as ne
+from natsort import natsort_keygen
 from sklearn.neighbors import NearestNeighbors
 from pciSeq.src.cell_call.log_config import logger
 
 
 class Cells(object):
     # Get rid of the properties where not necessary!!
     def __init__(self, _cells_df, config):
@@ -114,14 +115,16 @@
         # out['area_factor'] = np.ones(CellAreaFactor.shape)
         # logger.info('Overriden CellAreaFactor = 1')
         out['rel_radius'] = relCellRadius
         out['area'] = np.append(np.nan, img_obj.area)
         out['x0'] = np.append(-sys.maxsize, img_obj.x0.values)
         out['y0'] = np.append(-sys.maxsize, img_obj.y0.values)
         out['cell_label'] = np.append(0, img_obj.label.values)
+        if 'old_label' in img_obj.columns:
+            out['cell_label_old'] = np.append(0, img_obj.old_label.values)
         # First cell is a dummy cell, a super neighbour (ie always a neighbour to any given cell)
         # and will be used to get all the misreads. It was given the label=0 and some very small
         # negative coords
 
         return out, meanCellRadius
 
 # ----------------------------------------Class: Genes--------------------------------------------------- #
@@ -171,14 +174,15 @@
         # change in the future.
         # These two attributes take up a lot of space on the disk:
         # _gamma_bar and _log_gamma_bar
         # FYI: https://realpython.com/python-pickle-module/
         attributes = self.__dict__.copy()
         del attributes['_gamma_bar']
         del attributes['_log_gamma_bar']
+        return attributes
 
     # -------- PROPERTIES -------- #
     @property
     def gamma_bar(self):
         return self._gamma_bar.astype(self.config['dtype'])
 
     @gamma_bar.setter
@@ -318,15 +322,14 @@
             ne.evaluate("log(beta)", out=logb)
             return scipy.special.psi(r) - logb
         else:
             return scipy.special.psi(r) - np.log(beta).astype(dtype)
 
 
 # ----------------------------------------Class: SingleCell--------------------------------------------------- #
-# ----------------------------------------Class: SingleCell--------------------------------------------------- #
 class SingleCell(object):
     def __init__(self, scdata: pd.DataFrame, genes: np.array, config):
         self.isMissing = None  # Will be set to False is single cell data are assumed known and given as an input
                                # otherwise, if they are unknown, this will be set to True and the algorithm will
                                # try to estimate them
         # self.raw_data = self._raw_data(scdata, genes)
         self.config = config
@@ -389,15 +392,17 @@
         :return:
         """
         out = df.loc[:, (df != 0).any(axis=0)]
         return out
 
     def _helper(self, expr):
         # order by column name
-        expr = expr.copy().sort_index(axis=0).sort_index(axis=1, key=lambda x: x.str.lower())
+        # expr = expr.copy().sort_index(axis=0).sort_index(axis=1, key=lambda x: x.str.lower())
+        # expr = expr.copy().sort_index(axis=0).sort_index(axis=1, key=natsort_keygen())
+        expr = expr.copy().sort_index(axis=0).sort_index(axis=1, key=natsort_keygen(key=lambda y: y.str.lower()))
 
         # append at the end the Zero class
         expr['Zero'] = np.zeros([expr.shape[0], 1])
 
         # expr = self.config['Inefficiency'] * arr
         me = expr.rename_axis('gene_name').rename_axis("class_name", axis="columns")  # mean expression
         me = me + self.config['SpotReg']  # add the regularization parameter
```

## pciSeq/src/cell_call/main.py

```diff
@@ -1,8 +1,10 @@
 import numpy as np
+import pandas as pd
+import scipy.spatial as spatial
 import numpy_groupies as npg
 from pciSeq.src.cell_call.datatypes import Cells, Spots, Genes, SingleCell, CellType
 from pciSeq.src.cell_call.summary import collect_data
 import pciSeq.src.cell_call.utils as utils
 from pciSeq.src.cell_call.log_config import logger
 
 
@@ -10,14 +12,15 @@
     def __init__(self, _cells_df, _spots_df, scRNAseq, config):
         self.config = config
         self.cells = Cells(_cells_df, config)
         self.spots = Spots(_spots_df, config)
         self.genes = Genes(self.spots)
         self.single_cell = SingleCell(scRNAseq, self.genes.gene_panel, self.config)
         self.cellTypes = CellType(self.single_cell)
+        self.cells.class_names = self.single_cell.classes
         self.nC = self.cells.nC                         # number of cells
         self.nG = self.genes.nG                         # number of genes
         self.nK = self.cellTypes.nK                     # number of classes
         self.nS = self.spots.nS                         # number of spots
         self.nN = self.config['nNeighbors'] + 1         # number of closest nearby cells, candidates for being parent
                                                         # cell of any given spot. The last cell will be used for the
                                                         # misread spots. (ie cell at position nN is the background)
@@ -66,14 +69,15 @@
             self.has_converged, delta = utils.hasConverged(self.spots, p0, self.config['CellCallTolerance'])
             logger.info(' Iteration %d, mean prob change %f' % (i, delta))
 
             # replace p0 with the latest probabilities
             p0 = self.spots.parent_cell_prob
 
             if self.has_converged:
+                # self.counts_within_radius(20)
                 iss_df, gene_df = collect_data(self.cells, self.spots, self.genes, self.single_cell)
                 break
 
             if i == max_iter-1:
                 logger.info(' Loop exhausted. Exiting with convergence status: %s' % self.has_converged)
         return iss_df, gene_df
 
@@ -297,13 +301,66 @@
         # In this manner we will prevent the Zero class from being removed.
         mask[-1] = False
 
         # If a class size is smaller than 'min_class_size' then it will be assigned a weight of almost zero
         out[mask] = 10e-6
         self.cellTypes.alpha = out
 
+    def counts_within_radius(self, r):
+        """
+        calcs the gene counts within radius r of the centroid of each cell.
+        Units in radius are the same as in your coordinates x and y of your spots.
+        """
+
+        # check that the background (label=0) is at the top row
+        assert self.cells.ini_cell_props['cell_label'][0] == 0
+
+        # spots = pd.concat([self.spots.data, self.spots.data_excluded])
+        gene_names, gene_id = np.unique(self.spots.data.gene_name.values, return_inverse=True)  # that needs to be encapsulated!!! Make a function in the class to set the ids
+        spots = self.spots.data.assign(gene_id=gene_id)
+
+        xy_coords = spots[['x', 'y']].values
+        point_tree = spatial.cKDTree(xy_coords)
+        nearby_spots = point_tree.query_ball_point(self.cells.centroid, r)
+
+        out = np.zeros([self.cells.centroid.shape[0], len(gene_names)])
+
+        for i, d in enumerate(nearby_spots):
+            t = spots.gene_id[d]
+            b = np.bincount(t)
+            if len(gene_names) - len(b) < 0:
+                print('oops')
+            out[i, :] = np.pad(b, (0, len(gene_names) - len(b)), 'constant')
+
+        # for each cell get the most likely cell type
+        cell_type = []
+        for i, d in enumerate(self.cells.classProb):
+            j = self.cells.class_names[np.argmax(d)]
+            cell_type.append(j)
+
+        temp = pd.DataFrame({
+            'cell_label': self.cells.ini_cell_props['cell_label'],
+            'cell_type': cell_type
+        })
+
+        # if labels have been reassigned, there should be a column 'cell_label_old'.
+        # Relabelling will happen if for example the original labelling that was
+        # assigned from segmentation is not a continuous sequence of integers
+        # Append the original labels to the temp folder so there is a mapping between
+        # old and new labels
+        if 'cell_label_old' in self.cells.ini_cell_props.keys():
+            temp['cell_label_old'] = self.cells.ini_cell_props['cell_label_old']
+            temp = temp[['cell_label', 'cell_label_old', 'cell_type']]
+
+        # assert np.all(temp.cell_label==out.index)
+        df = pd.DataFrame(out, index=self.cells.centroid.index, columns=gene_names)
+        df = pd.merge(temp, df, right_index=True, left_on='cell_label', how='left')
+
+        # ignore the first row, it is the background
+        return df.iloc[1:, :]
+
```

## pciSeq/src/preprocess/spot_labels.py

```diff
@@ -42,28 +42,33 @@
 
 
 def reorder_labels(coo):
     """
     rearranges the labels so that they are a sequence of integers
     """
     label_image = coo.toarray()
-    _, idx = np.unique(label_image.flatten(), return_inverse=True)
-    return coo_matrix(idx.reshape(label_image.shape))
+    flat_arr = label_image.flatten()
+    u, idx = np.unique(flat_arr, return_inverse=True)
+
+    label_map = pd.DataFrame(set(zip(flat_arr, idx)), columns=['old_label', 'new_label'])
+    label_map = label_map.sort_values(by='old_label', ignore_index=True)
+    return coo_matrix(idx.reshape(label_image.shape)), label_map
 
 
 def stage_data(spots: pd.DataFrame, coo: coo_matrix) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
     """
     Reads the spots and the label image that are passed in and calculates which cell (if any) encircles any
     given spot within its boundaries. It also retrieves the coordinates of the cell boundaries, the cell
     centroids and the cell area
     """
 
+    label_map = None
     if coo.data.max() != len(set(coo.data)):
         logger.info(' The labels in the label image do not seem to be a sequence of successive integers. Relabelling the label image.')
-        coo = reorder_labels(coo)
+        coo, label_map = reorder_labels(coo)
 
     logger.info(' Number of spots passed-in: %d' % spots.shape[0])
     logger.info(' Number of segmented cells: %d' % len(set(coo.data)))
     logger.info(' Segmentation array implies that image has width: %dpx and height: %dpx' % (coo.shape[1], coo.shape[0]))
     mask_x = (spots.x >= 0) & (spots.x <= coo.shape[1])
     mask_y = (spots.y >= 0) & (spots.y <= coo.shape[0])
     spots = spots[mask_x & mask_y]
@@ -73,26 +78,32 @@
     spots = spots.assign(label=inc)
 
     # 2. Get cell centroids and area
     props = skmeas.regionprops(coo.toarray().astype(np.int32))
     props_df = pd.DataFrame(data=[(d.label, d.area, d.centroid[1], d.centroid[0]) for d in props],
                       columns=['label', 'area', 'x_cell', 'y_cell'])
 
+    # if there is a label map, attach it to the cell props.
+    if label_map is not None:
+        props_df = pd.merge(props_df, label_map, left_on='label', right_on='new_label', how='left')
+        props_df = props_df.drop(['new_label'], axis=1)
+
     # 3. Get the cell boundaries
     cell_boundaries = extract_borders_dip(coo.toarray().astype(np.uint32))
 
     assert props_df.shape[0] == cell_boundaries.shape[0] == np.unique(coo.data).shape[0]
     assert set(spots.label[spots.label > 0]) <= set(props_df.label)
 
     cells = props_df.merge(cell_boundaries)
     cells.sort_values(by=['label', 'x_cell', 'y_cell'])
     assert cells.shape[0] == cell_boundaries.shape[0] == props_df.shape[0]
 
     # join spots and cells on the cell label so you can get the x,y coords of the cell for any given spot
     spots = spots.merge(cells, how='left', on=['label'])
 
-    _cells = cells[['label', 'area', 'x_cell', 'y_cell']].rename(columns={'x_cell': 'x0', 'y_cell': 'y0'})
+    _cells = cells.drop(columns=['coords'])
+    _cells = _cells.rename(columns={'x_cell': 'x0', 'y_cell': 'y0'})
     _cell_boundaries = cells[['label', 'coords']].rename(columns={'label': 'cell_id'})
     _spots = spots[['x', 'y', 'label', 'Gene', 'x_cell', 'y_cell']].rename(columns={'Gene': 'target', 'x': 'x_global', 'y': 'y_global'})
 
     return _cells, _cell_boundaries, _spots
```

## Comparing `pciSeq-0.0.50.dist-info/LICENCE` & `pciSeq-0.0.51.dist-info/LICENCE`

 * *Files identical despite different names*

## Comparing `pciSeq-0.0.50.dist-info/METADATA` & `pciSeq-0.0.51.dist-info/METADATA`

 * *Files 5% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: pciSeq
-Version: 0.0.50
+Version: 0.0.51
 Summary: Probabilistic cell typing for spatial transcriptomics
 Home-page: https://github.com/acycliq/pciSeq
 Author: Dimitris Nicoloutsopoulos
 Author-email: dimitris.nicoloutsopoulos@gmail.com
 License: BSD
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: BSD License
@@ -17,14 +17,15 @@
 Requires-Dist: scikit-image
 Requires-Dist: scikit-learn
 Requires-Dist: tqdm
 Requires-Dist: flask
 Requires-Dist: numexpr
 Requires-Dist: diplib
 Requires-Dist: pyvips
+Requires-Dist: natsort
 Provides-Extra: interactive
 Requires-Dist: matplotlib (>=2.2.0) ; extra == 'interactive'
 Requires-Dist: jupyter ; extra == 'interactive'
 
 # pciSeq: Probabilistic Cell typing by In situ Sequencing
 A Python package that implements the cell calling algorithm as described in [Qian, X., et al. Nature Methods (2020)](https://www.nature.com/articles/s41592-019-0631-4)
 <p align="center">
@@ -49,37 +50,43 @@
 or, if you open the project in your IDE, then in your project settings, switch your interpreter to the interpreter of the `pciSeq` env. 
 ## Usage
 You need to create two `pandas dataframes` for the spots and the single cell data and a `coo_matrix` for the label image (which in 
 most cases will be the output of some image segmentation application). Then you pass them into the `pciSeq.fit()` method as follows: 
 ```
 import pciSeq
 
-res = pciSeq.fit(spots_df, label_image, scRNA_df)
+res = pciSeq.fit(spots=spots_df, coo=label_image, scRNAseq=scRNA_df)
 ```
 See the demo below for a more detailed explanation about the arguments of  `pciSeq.fit()` and its return values.
 
 There is also a fourth argument (optional) to override the default hyperparameter values which are initialised 
 by the [config.py](https://github.com/acycliq/pciSeq/blob/master/pciSeq/config.py) module. To pass user-defined hyperparameter values, create a `dictionary` with `keys` the
 hyperparameter names and `values` their new values. For example, to exclude all Npy and Vip spots you can do:
 
 ```
 import pciSeq
 
 opts = { 'exclude_genes': ['Npy', 'Vip'] }
-res = pciSeq.fit(spots_df, label_image, scRNAseq=scRNA_df, opts=opts)
+res = pciSeq.fit(spots=spots_df, coo=label_image, scRNAseq=scRNA_df, opts=opts)
 ```
 
 ## Demo
 You can run a pciSeq demo in google colab: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/acycliq/pciSeq/blob/master/notebooks/1_pciSeq.ipynb)
 
 ## Viewer
 An interactive viewer to explore the data runs on this [url](https://acycliq.github.io/visage/). Instructions about 
 building this viewer with your own data are [here](https://github.com/acycliq/visage). \
 If you have `v 0.0.49` or greater you can also launch the viewer automatically by 
 setting `opts = {'launch_viewer': True}` and passing it to `pciSeq.fit()`, see [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/acycliq/pciSeq/blob/master/notebooks/2_viewer.ipynb)
 
+## Change Log
+### [0.0.50] - 2023-05-27
+ - Single cell data are optional, more info can be found here [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/acycliq/pciSeq/blob/master/notebooks/3_pciSeq_without_singleCell_data.ipynb)
+
+ - `pciSeq.fit()` takes keyword arguments
+
 
 ## References 
 Qian, X., et al. (2020). Probabilistic cell typing enables fine mapping of closely related cell types in situ. Nat
 Methods 17, 101 - 106.
```

## Comparing `pciSeq-0.0.50.dist-info/RECORD` & `pciSeq-0.0.51.dist-info/RECORD`

 * *Files 4% similar despite different names*

```diff
@@ -1,23 +1,23 @@
 pciSeq/__init__.py,sha256=fTmu6nn7JN8jNECjrJBPpSdFJFz2BBO8et5igUincVY,2086
-pciSeq/app.py,sha256=DGmK8PXbNxJGsRIpl9Czx6cgXpMaaguP3-YyUe2GfOM,9326
+pciSeq/app.py,sha256=acedCJH1VSNzaTdZ4j8MykkTqSfAGiXdPKGyqmbcqsQ,9252
 pciSeq/config.py,sha256=M7iR2OAbptE5UldUa6t2cCiR9cyDwFEaTQbsKEe7yYY,3333
 pciSeq/make_plot.py,sha256=2AkncrVVvytr2keK505oMC0LV1Txli9WpsSGpjmht_4,3678
 pciSeq/src/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-pciSeq/src/_version.py,sha256=5EpyNnUNHdR6H711I3BZSKvH8qRHZCfaQSj21Qkn1t4,22
+pciSeq/src/_version.py,sha256=knhyYA07yyiNby_XiDk6x5W8EtHrJQtJI1DUCiTZ9jI,22
 pciSeq/src/cell_call/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-pciSeq/src/cell_call/datatypes.py,sha256=cZi8Wsr3KIkpQOiJXsKmxXuUFChPbyQGOu6AnNOS8CU,20581
+pciSeq/src/cell_call/datatypes.py,sha256=WxWGXn9rQ9pgvlBJtbl2RP2jIK4iEAKtw92AeyovNhk,20857
 pciSeq/src/cell_call/log_config.py,sha256=vnm9TxlMEVG_GLPfl21GDEWTafEk3B-drTLsVQMfFEE,2995
-pciSeq/src/cell_call/main.py,sha256=q_YGxPsVgLQW4fhRC5To27k3_HPYR6wI0cz7198j-4o,13927
+pciSeq/src/cell_call/main.py,sha256=n7DqQSmZvNRzYOzY2CYA_TexZjnSHuL9rH3Hj6MOFZ8,16544
 pciSeq/src/cell_call/summary.py,sha256=f2snLqJIeywtYuHrJLM-hRvifgqRU2u2aCKE_JrZJPs,3104
 pciSeq/src/cell_call/utils.py,sha256=sIrlkwdUgbfH7MmpOLw-Po4mTjsEWwNHHciUArHPN0A,8937
 pciSeq/src/preprocess/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
 pciSeq/src/preprocess/cell_borders.py,sha256=eTuFuexqGMFEV4upuTRInoaAi75NN46TIcvJ36TbZ4Q,11253
 pciSeq/src/preprocess/segmentation.py,sha256=9fc47EsxeFQnNtbJ0vm-RqATsBBMP73J5chdRFJocd4,6973
-pciSeq/src/preprocess/spot_labels.py,sha256=Y33S1wvhkgMKk9YjIOJZTDX5OyHpJDieM2exW0IrDB0,4090
+pciSeq/src/preprocess/spot_labels.py,sha256=ukYTnp0xnu9J4-bKCTLEDhHAfaQNCvE7teEFyU0OLRk,4585
 pciSeq/src/preprocess/utils.py,sha256=7UB-fOKsSY_a4MVPs7E6BoNZcwITqm2N2JFMZqNYbj0,88
 pciSeq/src/viewer/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
 pciSeq/src/viewer/run_flask.py,sha256=9HS_RW-O9xddAeiApd4m_w2JJmbqUGZ5lhQ-z_HUeVA,2256
 pciSeq/src/viewer/stage_image.py,sha256=ksrCt0LklOehdUdXTENFH3kvhSafRAtHAVHd22lSQQM,4511
 pciSeq/src/viewer/utils.py,sha256=8TtwWtoqox6Cy0otCIs6d10xwWUcEQgBkgVhG3HWVOg,7638
 pciSeq/static/2D/index.html,sha256=9aMR5hOaXugJpSOt3ypUY24z6blsfOS83Cd9LHbAHa0,10895
 pciSeq/static/2D/viewer/genes_datatable.html,sha256=v10lTjkXYeAv0DVuaWHOB6XgG887k-_ABXAgobDV-tY,6506
@@ -55,13 +55,13 @@
 pciSeq/static/2D/viewer/js/lib/js/leaflet-pip.js,sha256=QqsilU0SHTaKA_qNSv_jEDH53cWO6T_RdforfybnbKM,15374
 pciSeq/static/2D/viewer/js/lib/js/leaflet.textpath.js,sha256=i5cmMPqpqXShZz0TvXU3ZQUi2Y07SIbIpMYbhPE8gWE,5927
 pciSeq/static/2D/viewer/js/lib/js/preloader.js,sha256=eaFWRf9yT-Td4C2t01P8fDEZYszYTaGDgQ1vjevNIDI,265
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/MarkerContainer.js,sha256=Z8TADCqYEQr25OEF7wljsOVAOpx87UrjFPI2nvIty70,9097
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/bezier-easing.js,sha256=yMoRDqJiglg0M0eWM25WGA5zzLWuJpOyC3a8aGFx0jw,4434
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js,sha256=bZEnK8XSnShrpLYtOS7Xq7wnp-nW_nMZh4aaDc1Gt6U,613808
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js,sha256=A027Ub-oWXvKbz2B3pgMFS5xApmnEHXz9GQh79YZZnU,91874
-pciSeq-0.0.50.dist-info/LICENCE,sha256=6kbiFSfobTZ7beWiKnHpN902HgBx-Jzgcme0SvKqhKY,1091
-pciSeq-0.0.50.dist-info/METADATA,sha256=ZqX9Ba36SW7QYEtrZIvs93d0jDoq5f9IdyA6q_U7VIM,3697
-pciSeq-0.0.50.dist-info/WHEEL,sha256=G16H4A3IeoQmnOrYV4ueZGKSjhipXx8zc8nu9FGlvMA,92
-pciSeq-0.0.50.dist-info/entry_points.txt,sha256=RGOkAd0S7rrTuEInj-jKu5B7630i4R3mK9v9vx6nqc8,48
-pciSeq-0.0.50.dist-info/top_level.txt,sha256=mSonMN0SaB51m5yLGm5hFE1Fthf-2U4sRKxSezY-JMk,7
-pciSeq-0.0.50.dist-info/RECORD,,
+pciSeq-0.0.51.dist-info/LICENCE,sha256=6kbiFSfobTZ7beWiKnHpN902HgBx-Jzgcme0SvKqhKY,1091
+pciSeq-0.0.51.dist-info/METADATA,sha256=vJRvgKyqyPNcqYR6JufAvZbMF0_398bbzhEm2B6yLQU,4090
+pciSeq-0.0.51.dist-info/WHEEL,sha256=G16H4A3IeoQmnOrYV4ueZGKSjhipXx8zc8nu9FGlvMA,92
+pciSeq-0.0.51.dist-info/entry_points.txt,sha256=RGOkAd0S7rrTuEInj-jKu5B7630i4R3mK9v9vx6nqc8,48
+pciSeq-0.0.51.dist-info/top_level.txt,sha256=mSonMN0SaB51m5yLGm5hFE1Fthf-2U4sRKxSezY-JMk,7
+pciSeq-0.0.51.dist-info/RECORD,,
```

